src/components/cylc/tree/TreeItem.vue

```diff
@@ -32,39 +32,50 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
       <!-- the node value -->
       <!-- TODO: revisit these values that can be replaced by constants later (and in other components too). -->
       <div :class="getNodeDataClass()" @click="nodeClicked" v-if="node.type === 'cyclepoint'">
-        <task
-         :status="node.node.state"
-         :isHeld="node.node.isHeld"
-         :progress=0
-        />
+        <CylcObject :id="node.node.id">
+          <task
+           :status="node.node.state"
+           :isHeld="node.node.isHeld"
+           :progress=0
+          />
+        </CylcObject>
```

Comment:
Basic premise:

If you've got something you wan't to be interactive, wrap it with a `<CylcObject>` and provide it the Cylc ID for the "thing" your element represents.

This wrapper handles context menus, pop-ups and mutations.

I did think about adding this functionality into the task/job components themselves which would make developing new views a little easier, however, this allows groups of components to be wrapped up together.

I'm not 100% happy with this approach. For example, what if you wanted to wrap an entire row in the tree view? You could do it, however, CylcObject and the Tree compoments would be fighting for interaction events.

Open to ideas.

Reply:
Also this really won't work with multiple selection etc.

Reply:
Now thinking that this should be available somehow as a function something along the lines of:

```
interact([ids], [posX, posY])
```

This would, for example, allow the Tree view to handle multiple-selection then call the interact function with the IDs of the selected elements and the x,y position of the cursor to create the tooltip at?

In Vue terms this could be:

* A directive, however, I don't know whether a directive can create the context-menu component.
* A mixin for the Cylc "views", however, I don't know whether a mixin can create the context-menu component either.
* A wrapper component that the Cylc "views" are placed inside. The interact method would be passed to the view which would in turn pass it down to whatever components require it.

Reply:
>Basic premise:
>
>If you've got something you wan't to be interactive, wrap it with a <CylcObject> and provide it the Cylc ID for the "thing" your element represents.
>
>This wrapper handles context menus, pop-ups and mutations.
>
>I did think about adding this functionality into the task/job components themselves which would make developing new views a little easier, however, this allows groups of components to be wrapped up together.

Sounds like a good approach, similar to the Vuetify `v-skeleton` component. It wraps the component(s) and adds it some new feature.

>I'm not 100% happy with this approach. For example, what if you wanted to wrap an entire row in the tree view? You could do it, however, CylcObject and the Tree components would be fighting for interaction events.

Hmmm, well thought. But I think you should be able to capture the `click` or other events in the `CylcObject`, while allowing the children components to receive the event too, if necessary.

>Also this really won't work with multiple selection etc. I think for that to work this functionality would have to be much more centralised. Perhaps this is the stepping stone.

If I understand correctly, the issue would be if we allowed users to select a cycle point and a family proxy at the same time. What would happen when the user click to display the mutations? If so, I think we can just disable the mutations pop up when multiple elements of different types are selected (might be tricky to do that, but doable I think).

>In Vue terms this could be:
>
>    A directive, however, I don't know whether a directive can create the context-menu component.

Hmm, that would be tricky. Not sure how/where I would start to implement this with a directive. I haven't used any custom directive yet, but from what I've used in other components, it's normally used to enable/disable a feature of a component. So the components (`TreeItem` for example) would have to have built-in support to mutations, and that directive would just enable/disable it.

I think the current approach keeps a better separation of concerns, but might be wrong.

>A mixin for the Cylc "views", however, I don't know whether a mixin can create the context-menu component either.

Yeah, I can't think of a clean way, without making each component aware of the context-menu, at least in the template section we would need to have some `if`/`else`s.

>A wrapper component that the Cylc "views" are placed inside. The interact method would be passed to the view which would in turn pass it down to whatever components require it.

That could work, but not sure if that'd work with every sub-component. Worth a try if you think that'd be better. At least I think that's the easiest of the three options you suggested.

But I liked the simple `CylcObject` wrapper :smiley: